Upgrade guide for v4 changes. (#890)
microsoftgraph/msgraph-sdk-dotnet: docs/upgrade-to-v4.md

## Breaking changes

To improve the development experience provided by the SDK, it was necessary to make the following breaking changes in addition to the changes listed in the migration guide below:

 * .NET Standard minimum version bumped from `netstandard1.3` to `netstandard2.0`
 * .NET Framework minimum version bumped from `net45` to `net461`
 * Replacing Newtonsoft.Json dependency with System.Text.Json for serialization/de-serialization
 * Upgrading Microsoft.Graph.Core dependency to version 2.0.0

## Upgrade guide for breaking changes

The following section lists out the breaking changes requiring code changes from SDK users.

### System.Text.Json replaces Newtonsoft.Json

Updating to the latest version of the library involves the migration from [Newtonsoft.Json](https://www.newtonsoft.com/json) to System.Text.Json. You can always read more about the differences between the two libraries [here](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-5-0#table-of-differences-between-newtonsoftjson-and-systemtextjson).

> From the experience of the user of the SDKs, these are the main changes to look out for when upgrading.

Reviewer: ```suggestion
From the SDK developer experience, these are the main changes to look out for when upgrading.
```

 * .NET Standard minimum version bumped from `netstandard1.3` to `netstandard2.0`
 * .NET Framework minimum version bumped from `net45` to `net461`
 * Replacing Newtonsoft.Json dependency with System.Text.Json for serialization/de-serialization
 * Upgrading Microsoft.Graph.Core dependency to version 2.0.0

## Upgrade guide for breaking changes

The following section lists out the breaking changes requiring code changes from SDK users.

### System.Text.Json replaces Newtonsoft.Json

Updating to the latest version of the library involves the migration from [Newtonsoft.Json](https://www.newtonsoft.com/json) to System.Text.Json. You can always read more about the differences between the two libraries [here](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-5-0#table-of-differences-between-newtonsoftjson-and-systemtextjson).

From the experience of the user of the SDKs, these are the main changes to look out for when upgrading.

1. Object types and function parameters using Newtonsoft's JToken are now replaced by System.Text.Json's JsonDocument.

For example, the code sample :-

> ```cs

Reviewer 2: if you want those code blocks to be "under" the bullet point, you need to indent with 2 spaces. Same thing for the "For example" and "would change to" texts

Author: Thanks! Resolved via https://github.com/microsoftgraph/msgraph-sdk-dotnet/pull/890/commits/38e8635bae1b231b3659d52a34c9662d8b282538

GraphServiceClient graphClient = new GraphServiceClient( authProvider );

int index = 5;

JToken values = JToken.Parse("[[1,2,3],[4,5,6]]");

await graphClient.Me.Drive.Items["{id}"].Workbook.Tables["{id|name}"].Rows
	.Add(index,values)
	.Request()
	.PostAsync();

> ```

Author: I already have an open PR on devX APi to address this here https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/439

You can find other relevant JsonElement methods [here](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement?view=net-5.0#methods).

3. System.Text.Json enforces stricter json standards than Newtonsoft (e.g. trailing commas and comments are not allowed). 

In the event a user would like to use this, you can check out the guide [here](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-invalid-json?pivots=dotnet-5-0) and use it to override the current default serializer options.

For example:-
```cs
// Add extra options
var options = new JsonSerializerOptions
{
    ReadCommentHandling = JsonCommentHandling.Skip,
    AllowTrailingCommas = true,
};

Serializer serializer = new Serializer(options);
IResponseHandler responseHandler = new ResponseHandler(serializer); // Our Response Handler with custom Serializer

User me = await graphClient.Me.Request()
                .WithResponseHandler(responseHandler)
                .GetAsync();
```

> ### Azure identity

Reviewer 2: ```suggestion
### Azure Identity
```

it's a branded name

The Microsoft Graph library now supports the use of TokenCredential classes in the Azure.Identity library through the new `TokenCredentialAuthProvider`.

You can read more about available Credential classes [here](https://docs.microsoft.com/en-us/dotnet/api/overview/azure/identity-readme#key-concepts) and this is encouraged to be used in place of the `Microsoft.Graph.Auth` package. 

> For example, rather than using the [Username/password provider](https://docs.microsoft.com/en-us/graph/sdks/choose-authentication-providers?tabs=CS#UsernamePasswordProvider) from the `Microsoft.Graph.Auth` package, one could use the [UsernamePasswordCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.usernamepasswordcredential?view=azure-dotnet) class from `Azure.Identity` as follows.

Author: Valid point. Will action and resolve.

* `UpdateResponseAsync(UpdatedObject: Entity) : GraphResponse<T>`
* `DeleteResponseAsync() : GraphResponse`  (no generic here) 

Essentially, using the fluent APIs of the service libraries should remain the same as follows
```cs
User me = await graphClient.Me.Request()
                .GetAsync();
```

However, should you choose to want access to response headers and status codes a user can now do this.

```cs
GraphResponse<User> userResponse = await graphClient.Me.Request()
                                        .GetResponseAsync();

// Get the status code
HttpStatusCode status = userResponse.StatusCode;
// Get the headers
HttpResponseHeaders headers = userResponse.HttpHeaders;
// Get the user object using inbuilt serializer
User me = await response.GetResponseObjectAsync();
``` 

If you choose you can even deserialize the response in your own custom way as follows.

1. Use a custom IResponseHandler
```cs
ISerializer serializer = new CustomSerializer(); // Custom Serializer
IResponseHandler responseHandler = new ResponseHandler(serializer); // Our Response Handler with custom Serializer
    
GraphResponse<User> graphResponse = await graphServiceClient.Me.Request()

>                                                     .WithResponseHandler(responseHandler) // customized yay!

Reviewer 2: ```suggestion
                                                    .WithResponseHandler(responseHandler) // customized response handler
```
